File: c/Java/app/src/main/java/com/example/qispace_sdk_java_demo/QeepActivity.java
package com.example.qispace_sdk_java_demo;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONException;
import org.json.JSONObject;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

/**
 * This page demonstrates a use of the QEEP API to encrypt and decrypt a message with a QiSpace retrieved Quantum Key (QK) and iv
 * Additionally, it shows the usage of the SEQUR API as a by-product of unpacking the QK and its IV
 *
 * Prerequisites:
 * - An enterprise device token which will be used for QK retrieval from QiSpace Enterprise REST API
 * - A subkey intended for SEQUR decoding the QK after its retrieval. Realistically, subkey would be initialized on device setup
 *
 * **Ideally, setting up the subkey from SequrActivity must be done before getting QK as QK decryption requires subkey to be initialized
 *
 * Logical Flow:
 * 1. Press 'Get QK' to create and unpack a Quantum Key from QiSpace Enterprise
 * 2. Enter a message into the text field
 * 3. Press 'Encrypt' to QEEP encrypt the text field message
 * 4. Press 'Decrypt' to QEEP decrypt the encrypted text field message
 */

public class QeepActivity extends Activity {

    // Note: Assume sub_key is given for this demo page. Realistically, it is initialized on device setup

    // Note: Currently, max message length is 64;
    private byte[] encrypted_message;
    private byte[] decrypted_message;

    private int messageLen = 0;

    private byte[] qk;
    private byte[] message_iv;      //Note: Must be 16 bytes

    // Initialize the wrapper classes used for decoding and encrypting/decrypting
    qeepJavaWrapper qeep_wrapper = new qeepJavaWrapper();
    sequrJavaWrapper sequr_wrapper = new sequrJavaWrapper();

    private String QK_URL = QiSpace_Info.QISPACE_ENTERPRISE_BASE_URL+"/api/v1/qk";

    private RequestQueue QK_rq;

    private boolean have_successfully_encrypted = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.qeep_layout);

        // Initialize UI elements
        EditText message_in = findViewById(R.id.message_in);
        TextView encrypt_result = findViewById(R.id.encrypt_message_result);
        TextView decrypt_result = findViewById(R.id.decrypted_message_result);

        QK_rq = Volley.newRequestQueue(this);

        /**
         * QiSpace REST API response consists of
         * a payload, a pre-encrypted qeep key (QK),
         * and an iv, a variable that diminishes data encryption repetition,
         * all in Base64 format
         */
        // Sets up HTTP body info
        HashMap<String, Object> params = new HashMap<String, Object>();
        params.put("key_length", Integer.valueOf(32));      // For this demo, key_length was set to 32
        params.put("label", "message dev1");

        // Define request object for QiSpace API Call to get the encrypted QK payload
        JsonObjectRequest qk_request = new JsonObjectRequest
                (Request.Method.POST, QK_URL, new JSONObject(params), new Response.Listener<JSONObject>(){
                    @RequiresApi(api = Build.VERSION_CODES.O)
                    @Override
                    public void onResponse(JSONObject response){
                        byte[] qk_iv;
                        byte[] encoded_QK;
                        try {
                            // Gets and converts the payload and iv from Base64 to byte array
                            encoded_QK = Base64.getDecoder().decode(
                                    response.get("payload").toString().getBytes(StandardCharsets.UTF_8)
                            );
                            qk_iv = Base64.getDecoder().decode(
                                    response.get("iv").toString().getBytes(StandardCharsets.UTF_8)
                            );

                            // SEQUR decodes payload to get actual QK
                            qk = new byte[encoded_QK.length];
                            int ret = sequr_wrapper.decode(MainActivity.sub_key, qk_iv, encoded_QK, qk);
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                }, new Response.ErrorListener(){
                    @Override
                    public void onErrorResponse(VolleyError error){
                        System.out.println("ERROR: "+error.toString());
                    }
                }){
            // Sets additional credentials to get payload
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError{
                HashMap<String, String> headers = new HashMap<String, String>();
                headers.put("Authorization", MainActivity.token);
                headers.put("Content-Type", "application/json");
                return headers;
            }
        };

        // BUTTON: Handles QK retrieval using Quantropi's QiSpace API
        findViewById(R.id.button_get_qk).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // Trigger QiSpace API call to get QK
                QK_rq.add(qk_request);
            }
        });

        // BUTTON: Handles the process of encrypting the message using QK(Qeep Key) and iv
        findViewById(R.id.button_encrypt).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                closeKeyboard();

                // Gets message from text field input
                byte[] message = message_in.getText().toString().getBytes(Charset.forName("UTF-8"));
                messageLen = message.length;
                encrypted_message = new byte[messageLen];
                decrypted_message = new byte[messageLen];

                // Generates fixed fixed IV for this demo's encryption/decryption purposes
                // Note: The IV must be 16 bytes and should be randomly generated per encrypt operation or per message
                message_iv = hexStringToByteArray("0102030405060708090a010203040506");

                // Check qk/message_iv initialized
                if(qk == null){
                    encrypt_result.setText("Need to retrieve qk before encrypting");
                    return;
                }

                // Encrypts message with QEEP wrapper class
                int ret = qeep_wrapper.encrypt(qk, message_iv, message, encrypted_message);

                // Handle UI display with error codes
                if(ret == 1){
                    encrypt_result.setText(byteArrayToString(encrypted_message));
                    have_successfully_encrypted = true;
                }
                else{
                    encrypt_result.setText("Error code "+ret+" occurred in encryption. Consult error.h for more information");
                }
            }
        });

        // BUTTON: Handles the decryption of the message using qk (qeep key) and iv
        findViewById(R.id.button_decrypt).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                closeKeyboard();
                // Check qk initialized
                if(qk == null){
                    decrypt_result.setText("Need to retrieve qk before decrypting");
                    return;
                }
                // Ensures encryption before decryption
                if(!have_successfully_encrypted){
                    decrypt_result.setText("Need to encrypt before decrypting");
                    return;
                }

                // Decrypts the encrypted message
                // Note: Message iv should be identical to one used during encryption
                int ret = qeep_wrapper.decrypt(qk, message_iv, encrypted_message, decrypted_message);

                // Handle UI display with error codes based on decrypt result
                if(ret == 1){
                    decrypt_result.setText(new String(decrypted_message, StandardCharsets.US_ASCII));
                }
                else{
                    decrypt_result.setText("Error code "+ret+" occurred in decryption. Consult error.h for more information");
                }
            }
        });

        // BUTTON: go back to landing page
        findViewById(R.id.button_qeep_back).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent qeepToLanding = new Intent(QeepActivity.this, MainActivity.class);
                startActivity(qeepToLanding);
            }
        });
    }

    // Private Debugging procedure to print contents of byte array
    private static void printArr(byte[] arr){
        for(int i = 0;i<arr.length;i++){
            System.out.print(arr[i]+" ");
        }
        System.out.println();
    }

    private void closeKeyboard(){
        View view = this.getCurrentFocus();
        if(view != null){
            InputMethodManager imm = (InputMethodManager)getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    // Helper function to convert hexString to byte array
    private static byte[] hexStringToByteArray(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }

    // Helper method for convenient formatting
    private static String byteArrayToString(byte[] in){
        String decode_string = "";
        for(int b = 0;b<in.length;b++)
            decode_string += String.format("%x", in[b]);
        return decode_string;
    }
}

File: Kotlin/app/src/main/kotlin/com/example/demoqispaceapp/qeepActivity.kt
package com.example.demoqispaceapp

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.text.style.TabStopSpan
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.annotation.Nullable
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.example.demoqispaceapp.MainActivity.Companion.decodeHex
import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * This page demonstrates a use of the QEEP API to encrypt and decrypt a message with a QiSpace retrieved Quantum Key (QK) and iv
 * Additionally, it shows the usage of the SEQUR API as a by-product of unpacking the QK and its IV
 *
 * Prerequisites:
 * - An enterprise device token which will be used for QK retrieval from QiSpace Enterprise REST API
 * - A subkey intended for SEQUR decoding the QK after its retrieval. Realistically, subkey would be initialized on device setup
 *
 * **Ideally, setting up the subkey from sequrActivity must be done before getting QK as QK decryption requires subkey to be initialized
 * **Normally, it is initialized on device setup
 *
 * Logical Flow:
 * 1. Press 'Get QK' to create and unpack a Quantum Key from QiSpace Enterprise
 * 2. Enter a message into the text field
 * 3. Press 'Encrypt' to QEEP encrypt the text field message
 * 4. Press 'Decrypt' to QEEP decrypt the encrypted text field message
 */
class qeepActivity : Activity() {

    private lateinit var QK_rq: RequestQueue

    private val QK_URL = QiSpace_Info.QISPACE_ENTERPRISE_BASE_URL+"/api/v1/qk"

    // Note: Currently, maximum message length is 64
    private lateinit var encrypted_message: ByteArray
    private lateinit var decrypted_message: ByteArray

    private var messageLen = 0

    // Initialize payload variables
    private lateinit var QK : ByteArray
    private lateinit var message_iv : ByteArray     // Note: Must be 16 bytes

    // Initialize the wrapper classes used for decoding and encrypting/decrypting
    private val qeep_wrapper = qeepKotlinWrapper()
    private val sequr_wrapper = sequrKotlinWrapper()

    private var have_successfully_encrypted = false;

    override fun onCreate(@Nullable savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.qeep_layout)

        // Initialize UI elements
        val message_in = findViewById<EditText>(R.id.message_in)
        val encrypt_result = findViewById<TextView>(R.id.encrypt_message_result)
        val decrypt_result = findViewById<TextView>(R.id.decrypted_message_result)

        QK_rq = Volley.newRequestQueue(this)

        /**
         *       QiSpace QK API response consists of
         *       a payload, a pre-encoded Qeep Key (QK),
         *       and an iv, a variable that diminishes data encryption repetition,
         *       all in Base64 format
         */
        // Sets up QiSpace API HTTP call body
        val params = JSONObject()
        params.put("key_length", 32)
        params.put("label", "demo-device")

        // QiSpace API Call to get the encrypted QK payload
        val qk_request : JsonObjectRequest = @RequiresApi(VERSION_CODES.O)
        object: JsonObjectRequest(Request.Method.POST, QK_URL, params, Response.Listener{ stuff ->
//            println(stuff)
            // Gets and converts the payload and iv from Base64 to byte array
            var encoded_QK : ByteArray = Base64.getDecoder().decode(
                stuff["payload"].toString().toByteArray(Charsets.UTF_8)
            )
            var qk_iv : ByteArray = Base64.getDecoder().decode(
                stuff["iv"].toString().toByteArray(Charsets.UTF_8)
            )

            QK = ByteArray(encoded_QK.size)
            // SEQUR decode the payload to get actual QK
            val ret = sequr_wrapper.decode(MainActivity.sub_key, qk_iv, encoded_QK, QK)

        }, Response.ErrorListener{ error ->
            println("Error: %s".format(error.toString()))
        }){

            // Sets HTTP request headers to get payload
            override fun getHeaders(): MutableMap<String, String>{
                val headers = HashMap<String, String>()
                headers["Authorization"] = MainActivity.token
                headers["Content-Type"] = "application/json"
                return headers
            }
        }

        // BUTTON: Handles QK retrieval using Quantropi's QiSpace API
        findViewById<Button>(R.id.button_get_qk).setOnClickListener{
            //Trigger QiSpace API call to get QK
            QK_rq.add(qk_request)
        }

        // BUTTON: Handles the process of encoding the message using QK(qeep key) and iv
        findViewById<Button>(R.id.button_encrypt).setOnClickListener {
            hideKeyboardFrom(this, message_in)

            // Get message from user input
            val message = message_in.text.toString().toByteArray(Charsets.UTF_8)
            messageLen = message.size
            encrypted_message = ByteArray(messageLen)
            decrypted_message = ByteArray(messageLen)

            // Generate random 16 bytes (MANDATORY) iv per encrypt operation or per message
            // The message iv is simply fixed for this demo
            message_iv = "0102030405060708090a010203040506".decodeHex()

            // Check qk/message_iv initialized
            if(!(this::QK.isInitialized)){
                encrypt_result.text = "Need to retrieve qk before decryption"
                return@setOnClickListener
            }

            // Encrypt the message with QEEP wrapper class
            val ret = qeep_wrapper.encrypt(QK, message_iv, message, encrypted_message)

            // Handle UI display based on error codes
            if(ret == 1){
                encrypt_result.text = byteArrayToString(encrypted_message)
                have_successfully_encrypted = true;
            }
            else{
                encrypt_result.text = "Error code "+ret+" occurred in encryption. Consult eror.h for more information";
            }
        }

        // BUTTON: Handles the decryption of the message using QK (qeep key) and iv
        findViewById<Button>(R.id.button_decrypt).setOnClickListener {
            hideKeyboardFrom(this, message_in)

            // Check QK & encrypted_message initialized
            if(!(this::QK.isInitialized)){
                decrypt_result.text = "Need to retrieve qk before decryption"
                return@setOnClickListener
            }
            if(!have_successfully_encrypted){
                decrypt_result.text = "No encrypted message available to decrypt"
                return@setOnClickListener
            }

            // Decrypt the encrypted message
            // Message iv should be identical to one used during encryption
            val ret2 = qeep_wrapper.decrypt(QK, message_iv, encrypted_message, decrypted_message)

            // Handle UI display based on error codes
            if(ret2 == 1){
                decrypt_result.text = String(decrypted_message, StandardCharsets.US_ASCII)
            }
            else{
                decrypt_result.text = "Error code "+ret2+" occurred during decryption. Consult error.h for more information"
            }
        }

        // BUTTON: go back to landing page
        findViewById<Button>(R.id.button_qeep_back).setOnClickListener{
            val qeepToLanding = Intent(this, MainActivity::class.java)
            startActivity(qeepToLanding)
        }
    }

    fun hideKeyboardFrom(context: Context, view: View) {
        val imm: InputMethodManager =
            context.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }

    // Helper function to convert hexString to byte array
    fun String.decodeHex(): ByteArray {
        check(length % 2 == 0) { "Must have an even length" }
        return chunked(2)
            .map { it.toInt(16).toByte() }
            .toByteArray()
    }

    // Private debugging procedure to print contents of byte array
    private fun printArr(arr: ByteArray) {
        for (i in arr.indices) {
            print(arr[i].toString() + " ")
        }
        println()
    }

    // Helper method for convenient formatting
    private fun byteArrayToString(input : ByteArray) : String{
        var decode_string : String = ""
        for(b in 0..input.size-1){
            decode_string += String.format("%x", input[b])
        }
        return decode_string
    }
}
